qualtran/bloqs/chemistry/sf/prepare.py

```diff
@@ -0,0 +1,178 @@
+
+    Prepares the state in Eq. B11 of Ref [1].
+
+    Currently we only provide costs as listed in Ref[1] without their corresponding decompositions.
+
+    Args:
+        num_aux: Dimension of auxiliary index for single factorized Hamiltonian. Call L in Ref[1].
+        num_spin_orb: The number of spin orbitals. Typically called N.
+        num_bits_state_prep: The number of bits of precision for coherent alias
+            sampling. Called aleph (fancy N) in Ref[1].
+        num_bits_rot_aa: Number of bits of precision for rotations for amplitude
+            amplification in uniform state preparation. Called b_r in Ref[1].
+        adjoint: Whether this bloq is daggered or not. This affects the QROM cost.
+
+    Registers:
+        l: register to store L values for auxiliary index.
+        p: spatial orbital index. range(0, num_spin_orb // 2)
+        q: spatial orbital index. range(0, num_spin_orb // 2)
+        succ_pq: flag for success of this state preparation.
+
+    Refererences:
+        [Even More Efficient Quantum Computations of Chemistry Through Tensor
+            Hypercontraction](https://arxiv.org/abs/2011.03494) Appendix B. Listing 2, page 44.
+        [Qubitization of Arbitrary Basis Quantum Chemistry Leveraging Sparsity
+        and Low Rank Factorization](https://quantum-journal.org/papers/q-2019-12-02-208/) Sec. 3.2
+    """
+
+    num_aux: int
+    num_spin_orb: int
+    num_bits_state_prep: int
+    num_bits_rot_aa: int = 8
+    adjoint: bool = False
+    kp1: int = 1
+    kp2: int = 1
+
+    def pretty_name(self) -> str:
+        dag = '†' if self.adjoint else ''
+        return f"In-Prep{dag}"
+
+    @cached_property
+    def signature(self) -> Signature:
+        n = (self.num_spin_orb // 2 - 1).bit_length()
+        return Signature.build(l=self.num_aux.bit_length(), p=n, q=n, succ_pq=1)
+
+    def build_call_graph(self, ssa: 'SympySymbolAllocator') -> Set['BloqCountT']:
+        n = (self.num_spin_orb // 2 - 1).bit_length()
+        # prep uniform upper triangular.
+        cost_a = (Toffoli(), 6 * n + 2 * self.num_bits_rot_aa - 7)
+        # contiguous index
+        cost_b = (Toffoli(), n**2 + n - 1)
+        # Note the data size is for storing the upper triangular matrices of
+        # size N/2, so N/2 (N/2 + 1)/2. There is an error in the reference
+        # equation B13 of Reg[1] (it is correct in Ref[2], and openfermion).
+        # QROAM
+        cost_c = (
+            QROAMTwoRegs(
+                self.num_aux + 1,
+                self.num_spin_orb**2 // 8 + self.num_spin_orb // 4,
+                self.kp1,
+                self.kp2,
+                2 * n + self.num_bits_state_prep + 2,
+                adjoint=self.adjoint,
+            ),
+            1,
+        )
+        # inequality test
+        cost_d = (Toffoli(), self.num_bits_state_prep)
+        # controlled swap
+        cost_e = (Toffoli(), 2 * n)
+        return {cost_a, cost_b, cost_c, cost_d, cost_e}
```

Comment:
provide better names for `cost_{a,b,c,d,e}`. presumably these should each be replaced with bloqs in the future, yes? Don't we have bloqs with costs for inequality testing and controlled swap?

Reply:
For the call graphs I was thinking to use the costs as provided exactly by the references so in some cases I would use Toffoli costs rather than the bloq as the inequality tests sometimes are inverted with zero Toffoli cost. In the decomposition I would then use the appropriate bloq and we could track the disagreement over time. BUT this isn't the case here and I should have used bloqs.

Reply:
The a/b/c/d notation was in keeping with the listings in the paper, but I can make it more descriptive.

Reply:
As a matter of style, I'd use descriptive variable names in code and use the comments to note the corresponence with the a/b/c/d notation in the paper :)